File: 00525_Array__Contiguous_Array/sol2.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {

        unordered_map<int, int> _seen;  // <sum, 1st idx of nums>

        int ans = 0;
        int sum = 0;
        for (int i = 0; i < nums.size(); i++) {
            sum += (nums[i] == 0 ? -1 : 1);

            if (sum == 0) {
                ans = max(ans, i + 1);
                continue;
            }

            if (_seen.count(sum)) {
                ans = max(ans, i - _seen[sum]);
            } else {
                _seen[sum] = i;
            }
        }
        return ans;
    }
};

File: 00032_Stack__Longest_Valid_Parentheses/sol2.py
class Solution:

    def longestValidParentheses(self, s: str) -> int:

        if len(s) < 2:
            return 0

        stack = []
        flag = [0] * len(s)

        for idx, ch in enumerate(s):
            if len(stack) == 0:
                stack.append((idx, ch))
                continue
            if stack[-1][1] == '(' and ch == ')':
                flag[stack[-1][0]] = 1
                flag[idx] = 1
                stack = stack[:-1]
            else:
                stack.append((idx, ch))

        cnt = 0
        ans = 0
        for i in flag:
            if i == 1:
                cnt += 1
            else:
                ans = max(ans, cnt)
                cnt = 0
        ans = max(ans, cnt)
        
        return  ans   
        
